core/src/main/scala/kafka/server/KafkaApis.scala

```
@@ -1322,6 +1326,30 @@ class KafkaApis(val requestChannel: RequestChannel,
     txnCoordinator.handleInitPid(initPidRequest.transactionalId, sendResponseCallback)
   }
 
+  def handleEndTxnRequest(request: Request): Unit = {
+    throw new UnsupportedOperationException("Not implemented yet.")
+  }
+
+  def handleAbortTxnRequest(request: Request): Unit = {
+    throw new UnsupportedOperationException("Not implemented yet.")
+  }
+
+  def handleAddPartitionToTxnRequest(request: Request): Unit = {
+    throw new UnsupportedOperationException("Not implemented yet.")
+  }
+
+  def handleAddOffsetsToTxnRequest(request: Request): Unit = {
+    throw new UnsupportedOperationException("Not implemented yet.")
+  }
+
+  def handleWriteTxnMarkerRequest(request: Request): Unit = {
+    throw new UnsupportedOperationException("Not implemented yet.")
+  }
+
+  def handleTxnOffsetCommitRequest(request: Request): Unit = {
+    throw new UnsupportedOperationException("Not implemented yet.")
```

Could we send dummy responses instead of throwing an exception? KAFKA-4954 is trying to add a mechanism to delay the response of every client request due to request level throttling. It would be harder to do that if the implementation just throws an exception.

---

I think the idea was not to expose these new protocol APIs until we have had a chance to use them to verify that it's all good. How does this make it harder for KAFKA-4954? There are tests that exercise all `ApiKeys`?

---

In KAFKA-4954, each request handler call is changed to the following before a response is sent. It's much easier to add such logic if the request handler sends a response instead of throwing an exception. Otherwise, anyone who removes the UnsupportedOperationException has to remember adding that logic later on.

```
      def createResponse(throttleTimeMs: Int): AbstractResponse = new UpdateMetadataResponse(Errors.CLUSTER_AUTHORIZATION_FAILED)
     sendResponseMaybeThrottle(request, createResponse)
```

---

We can return an error. But the pattern you outlined works fine with exceptions too.

```scala
def createResponse(throttleTimeMs: Int: AbstractResponse = throw new UnsupportedOperationException("Not yet implemented")
sendResponseMaybeThrottle(request, createResponse)
```